rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/util/Diagnostics.java

```diff
@@ -91,71 +89,120 @@ public static interface Template {
 
-        // Highlight selected parts of the error tree
-        if (ERROR_LOCATION_HIGHLIGHT != null) {
-            // Just the error location
-            ISourceLocation errorLoc = factory.sourceLocation(skippedLoc,
-                    skippedLoc.getOffset(), 1,
-                    skippedLoc.getBeginLine(), skippedLoc.getBeginLine(),
-                    skippedLoc.getBeginColumn(), skippedLoc.getBeginColumn() + 1);
-            diagnostics.add(cm -> new Diagnostic(toRange(errorLoc, cm), "Recovered parse error location",
-                    ERROR_LOCATION_HIGHLIGHT, "parser"));
-        }
+        // points at where we had to start skipping. This is typically the same error
+        // location as if no error recovery was done.
+        ISourceLocation stuckLoc = factory.sourceLocation(
+            skippedLoc.top(),
+            skippedLoc.getOffset(),
+            1,
+            skippedLoc.getBeginLine(),
+            skippedLoc.getBeginLine(),
+            skippedLoc.getBeginColumn(),
+            skippedLoc.getBeginColumn() + 1);
+
+        // this spans the recognized prefix until the skipped part
+        ISourceLocation prefixLoc = factory.sourceLocation(
+                completeErrorTreeLoc.top(),
+                completeErrorTreeLoc.getOffset(),
+                skippedLoc.getOffset() - completeErrorTreeLoc.getOffset(),
+                completeErrorTreeLoc.getBeginLine(),
+                skippedLoc.getBeginLine(),
+                completeErrorTreeLoc.getBeginColumn(),
+                skippedLoc.getBeginColumn()
+        );
+
+        String nonterminal = SymbolAdapter.toString(TreeAdapter.getType(errorTree), false);
 
-        if (ERROR_TREE_HIGHLIGHT != null) {
-            // The whole error tree
-            diagnostics.add(cm -> new Diagnostic(toRange(errorTreeLoc, cm), "Recovered parse error", ERROR_TREE_HIGHLIGHT, "parser"));
-        }
+        List<Template> diagnostics = new ArrayList<>();
 
-        if (PREFIX_HIGHLIGHT != null) {
-            // The recognized prefix
-            int prefixLength = skippedLoc.getOffset()-errorTreeLoc.getOffset();
-            if (prefixLength > 0) {
-                ISourceLocation prefixLoc = factory.sourceLocation(errorTreeLoc,
-                        errorTreeLoc.getOffset(), skippedLoc.getOffset()-errorTreeLoc.getOffset(),
-                        errorTreeLoc.getBeginLine(), skippedLoc.getBeginLine(),
-                        errorTreeLoc.getBeginColumn(), skippedLoc.getBeginColumn());
-                diagnostics.add(cm -> new Diagnostic(toRange(prefixLoc, cm), "Recovered parse error prefix", PREFIX_HIGHLIGHT, "parser"));
+        diagnostics.add(cm -> {
+            var d = new Diagnostic(
+                toRange(stuckLoc, cm),
+                "Parsing got stuck here",
+                DiagnosticSeverity.Error,
+                "parser");
+
+            List<DiagnosticRelatedInformation> related = new ArrayList<>();
+
+            related.add(related(cm, stuckLoc,
```

Comment:
can you show a screenshot/video how this looks for users?

Reply:
This is the folded diagnostic without any "related information":

<img width="545" height="37" alt="image" src="https://github.com/user-attachments/assets/02d33499-5f2d-4bb3-a7fe-ea8f5fc5e60d" />

This is with the unfolded unformation (currently only for non-experts):

<img width="760" height="113" alt="image" src="https://github.com/user-attachments/assets/2c8a5b4f-7c55-4ffd-8dbd-a5d3432d7d6e" />

Every line links to a different part of the source code. Even though it may start at the same position.

Reply:
This is what the error hover looks like:

<img width="876" height="347" alt="image" src="https://github.com/user-attachments/assets/00e65ac3-b50a-4488-bb2e-8811ee6994d8" />